### PR TITLE
Implicitly close related portals if a prepared statement is closed

### LINKED ISSUE #8782: Second SELECT request failed while using ODBC connection
**CrateDB version**: 3.3.3

**OS version**: 
>> lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic

**Used PostgreSQL ODBC driver**:
>> apt show odbc-postgresql
Package: odbc-postgresql
Version: 1:11.01.0000-1.pgdg18.04+1

**odbc.ini / odbcinst.ini**
>> cat /etc/odbc.ini
[crate]
Driver = PostgreSQL ANSI
Host = crate00
Server = crate00
ServerName = crate00
Database = doc
Port = 5432
Trace=yes
TraceFile=/tmp/odbc.log

>> cat /etc/odbcinst.ini
[PostgreSQL ANSI]
Description=PostgreSQL ODBC driver (ANSI version)
Driver=psqlodbca.so
Setup=libodbcpsqlS.so
Debug=0
CommLog=1
UsageCount=1
>>
>>[PostgreSQL Unicode]
Description=PostgreSQL ODBC driver (Unicode version)
Driver=psqlodbcw.so
Setup=libodbcpsqlS.so
Debug=0
CommLog=1
UsageCount=1

**Problem description**:
I trying to connect to crate via ODBC.
I successfully connected to crate, and first SELECT successfully executes.
But second SELECT is always failed.
Odbc trace file and tcpdump in attachment.
Exception raised from `BatchPortal.java:199`

[odbc.log](https://github.com/crate/crate/files/3265217/odbc.log)
[odbc.pcap.txt](https://github.com/crate/crate/files/3265223/odbc.pcap.txt) (rename it into odbc.pcap before open)

**Steps to reproduce**:

`iodbctest "DSN=crate;UID=crate;"`
`iODBC Demonstration program`
`This program shows an interactive SQL processor`
`Driver Manager: 03.52.0812.0326`
`Driver: 11.01.0000 (psqlodbca.so)`
` `
`SQL>select 1+2 as x;`
` `
`x`
`--------------------`
`3`
` `
` result set 1 returned 1 rows.`
` `
` `
`SQL>select 2+3 as x;`
**`1: SQLExec = ERROR: Only write operations are allowed in Batch statements;`
`Error while executing the query (1) SQLSTATE=0A000`**
` `
`SQL>some invalid statement will raise error`
`1: SQLExec = ERROR: SQLParseException: line 1:1: mismatched input 'some' expecting {'SELECT', 'DEALLOCATE', 'CREATE', 'ALTER', 'KILL', 'BEGIN', 'COMMIT', 'EXPLAIN', 'SHOW', 'OPTIMIZE', 'REFRESH', 'RESTORE', 'DROP', 'INSERT', 'DELETE', 'UPDATE', 'SET', 'RESET', 'COPY', 'GRANT', 'DENY', 'REVOKE'};`
`Error while preparing parameters (1) SQLSTATE=XX000`
` `
`SQL>select 3+4 as x;`
` `
`x`
`--------------------`
`7`
` `
` result set 1 returned 1 rows.`
` `
` `
`SQL>select 4+5 as x;`
**`1: SQLExec = ERROR: Only write operations are allowed in Batch statements;`
`Error while executing the query (1) SQLSTATE=0A000`**
` `
`SQL>`

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

See commits for details. (Last commit contains the actual fix, third commit is a refactoring to make the fix simpler)

From the PostgreSQL wire protocol docs:

> Note that closing a prepared statement implicitly closes any open
> portals that were constructed from that statement.

We didn't do this so far, which caused the issue described in
#8782

This fixes #8782

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/action/sql/Session.java [hunk 1]
```diff
@@ -391,49 +388,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
...
-                clearState();
-                return portal.sync(planner, jobsLogs);
+                return singleExec(portal, deferredExecution.resultReceiver(), deferredExecution.maxRows());
+
             default:
-                throw new IllegalStateException(
-                    "Shouldn't have more than 1 pending execution. Got: " + pendingExecutions);
+                List<DeferredExecution> deferredExecutions = List.copyOf(this.deferredExecutions);
+                this.deferredExecutions.clear();
+                Map<Statement, List<DeferredExecution>> deferredExecutionsByStatement = deferredExecutions
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/action/sql/Session.java]: We could eagerly create the grouped structure and probably also mix singleExec and bulkExec within the same `sync` call. 

But I'd keep it for now like this for the backport and followup on this on master only.

File: sql/src/main/java/io/crate/action/sql/Session.java [hunk 2]
```diff
@@ -376,49 +387,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
...
+        var plannerContext = new PlannerContext(
+            clusterState,
+            routingProvider,
+            jobId,
+            executor.functions(),
+            txnCtx,
+            0
+        );
+
+        var entry = deferredExecutionsByStatement.entrySet().iterator().next();
```
Comment (andreidan) [inline sql/src/main/java/io/crate/action/sql/Session.java]: shall we add an assertion on the size of the `deferredExecutionsByStatement` as we have a hard expectation of what it should be?

File: sql/src/main/java/io/crate/action/sql/Session.java [hunk 3]
```diff
@@ -376,49 +387,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
...
+        List<CompletableFuture<?>> resultReceiverFutures = Lists2.map(toExec, x -> x.resultReceiver().completionFuture());
+        CompletableFuture<Void> allResultReceivers = CompletableFuture.allOf(resultReceiverFutures.toArray(new CompletableFuture[0]));
+
+        return allRowCounts
+            .exceptionally(t -> null) // swallow exception - failures are set per item in emitResults
+            .thenAccept(ignored -> emitResults(jobId, jobsLogs, toExec, rowCounts))
+            .runAfterBoth(allResultReceivers, () -> {});
+    }
+
+    private static void emitResults(UUID jobId,
```
Comment (andreidan) [inline sql/src/main/java/io/crate/action/sql/Session.java]: would `emitRowCountResults` be a more descriptive name for this?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/action/sql/Session.java]: Changed it to `emitRowCountsToResultReceivers`

File: sql/src/main/java/io/crate/action/sql/Session.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/action/sql/Session.java]: I changed the signature to take the ` Statement` + `List<DeferredExecutions>` diretly. This way the unpacking of the entrySet is right after the `size` check.

**Comment by mfussenegger** [comment]
retest this please

**Pull request merged**